From 0c207a7270223f45e7e432fe1603c1e0a5debd9f Mon Sep 17 00:00:00 2001
From: Pierre K <pierrekn@gmail.com>
Date: Wed, 17 Aug 2022 17:52:35 +0200
Subject: [PATCH 174/196] Fix build with Boost 1.77.0 (#263)

Summary:
BOOST_FILESYSTEM_C_STR changed to accept the path as an argument

This is a backport of [[https://github.com/bitcoin/bitcoin/pull/22713 | core#22713]]

Test Plan: `ninja all check-all`

Reviewers: #bitcoin_abc, Fabien

Reviewed By: #bitcoin_abc, Fabien

Differential Revision: https://reviews.bitcoinabc.org/D10769

Co-authored-by: Rafael Sadowski <rafael@sizeofvoid.org>
---
 src/fs.cpp                            | 5 +++++
 src/wallet/test/db_tests.cpp          | 4 ++++
 src/wallet/test/init_test_fixture.cpp | 4 ++++
 3 files changed, 13 insertions(+)

diff --git a/src/fs.cpp b/src/fs.cpp
index 752a442e1..25f01ba72 100644
--- a/src/fs.cpp
+++ b/src/fs.cpp
@@ -236,8 +236,13 @@ void ofstream::close() {
 }
 #else // __GLIBCXX__
 
+#if BOOST_VERSION >= 107700
+static_assert(
+    sizeof(*BOOST_FILESYSTEM_C_STR(fs::path())) == sizeof(wchar_t),
+#else
 static_assert(
     sizeof(*fs::path().BOOST_FILESYSTEM_C_STR) == sizeof(wchar_t),
+#endif // BOOST_VERSION >= 107700
     "Warning: This build is using boost::filesystem ofstream and ifstream "
     "implementations which will fail to open paths containing multibyte "
     "characters. You should delete this static_assert to ignore this warning, "
diff --git a/src/wallet/test/db_tests.cpp b/src/wallet/test/db_tests.cpp
index 48583f6d3..3afa02e37 100644
--- a/src/wallet/test/db_tests.cpp
+++ b/src/wallet/test/db_tests.cpp
@@ -18,7 +18,11 @@ BOOST_AUTO_TEST_CASE(getwalletenv_file) {
     std::string test_name = "test_name.dat";
     const fs::path datadir = GetDataDir();
     fs::path file_path = datadir / test_name;
+#if BOOST_VERSION >= 107700
+    std::ofstream f(BOOST_FILESYSTEM_C_STR(file_path));
+#else
     std::ofstream f(file_path.BOOST_FILESYSTEM_C_STR);
+#endif // BOOST_VERSION >= 107700
     f.close();
 
     std::string filename;
diff --git a/src/wallet/test/init_test_fixture.cpp b/src/wallet/test/init_test_fixture.cpp
index 7d083a0e5..774bbddfb 100644
--- a/src/wallet/test/init_test_fixture.cpp
+++ b/src/wallet/test/init_test_fixture.cpp
@@ -34,7 +34,11 @@ InitWalletDirTestingSetup::InitWalletDirTestingSetup(
     fs::create_directories(m_walletdir_path_cases["default"]);
     fs::create_directories(m_walletdir_path_cases["custom"]);
     fs::create_directories(m_walletdir_path_cases["relative"]);
+#if BOOST_VERSION >= 107700
+    std::ofstream f(BOOST_FILESYSTEM_C_STR(m_walletdir_path_cases["file"]));
+#else
     std::ofstream f(m_walletdir_path_cases["file"].BOOST_FILESYSTEM_C_STR);
+#endif // BOOST_VERSION >= 107700
     f.close();
 }
 
-- 
2.39.5 (Apple Git-154)

